kotlin/android/app/src/main/java/dev/firezone/android/tunnel/NetworkMonitor.kt

```diff
@@ -18,19 +18,21 @@ class NetworkMonitor(private val tunnelService: TunnelService) : ConnectivityMan
         linkProperties: LinkProperties,
     ) {
         Log.d("NetworkMonitor", "OnLinkPropertiesChanged: $network: $linkProperties")
-        if (tunnelService.tunnelState != TunnelService.Companion.State.UP) {
-            tunnelService.tunnelState = TunnelService.Companion.State.UP
-            tunnelService.updateStatusNotification(TunnelStatusNotification.Connected)
-        }
+        tunnelService.connlibSessionPtr?.let {
```

Comment:
@conectado I'm not 100% sure the crash is happening on the JVM side. From the backtrace it looks like it's in the set_dns() call in connlib. I think you mentioned you knew what the issue was here?

Reply:
What I imagined is that given that we can set the `connlibSessionPtr` while the `set_dns` in connlib is being called, it might be the case that the pointer is being cleaned by the gc at some point and we try to deref an invalid memory location.

It'd be useful to try and see if we can get the debug symbols for the backtrace.

Reply:
Unfortunately getting the debug symbols into Firebase is no trivial task. I've tried a few times and haven't been successful.

You're welcome to take a shot at it -- see the backtrace for this error in Firebase for example.

Is it possible this occurs when connlib shuts down?

Reply:
yeah, I think that's the case, basically the same race condition mentioned in this PR but within connlib

Reply:
Should we make the `connlibSessionPtr` a mutex? Even if that's not the bug it makes sense that we don't want multiple threads modifying the pointer at the same time.

Reply:
On the JVM side? I don't think that would be necessary since it's never written to -- it's simply a Long.

Could it be possible the session pointer is changing on the Rust side and the JVM is never notified?

If not then I think the only possibility is what this PR fixes -- the monitor runs in another thread from TunnelService and isn't necessarily stopped before disconnect()

Reply:
We don't change the pointer from connlib, save for freeing on `disconnect`, we use it as a reference to the session always.

I think a Mutex would be the proper solution, since you have 1 thread that's accessing it, the `NetworkMonitor` thread and other that could null it, the activity thread. It's not "just" a `Long` in the sense that part of the unsafe contract with Rust is that the pointer is valid. Furthermore, I think that simulateneously freeing memory in a concurrent thread is UB [ref](https://doc.rust-lang.org/std/ptr/index.html#safety):

> All accesses performed by functions in this module are non-atomic in the sense of [atomic operations](https://doc.rust-lang.org/std/sync/atomic/index.html) used to synchronize between threads. This means it is undefined behavior to perform two concurrent accesses to the same location from different threads unless both accesses only read from memory. Notice that this explicitly includes [read_volatile](https://doc.rust-lang.org/std/ptr/fn.read_volatile.html) and [write_volatile](https://doc.rust-lang.org/std/ptr/fn.write_volatile.html): Volatile accesses cannot be used for inter-thread synchronization

Reply:
> I think a Mutex would be the proper solution, since you have 1 thread that's accessing it, the NetworkMonitor thread and other that could null it, the activity thread.

This PR achieves that I think? Now that `onRevoke` is implemented I think this issue may actually go away. According to the Android docs for [`unregisterNetworkCallback`](https://developer.android.com/reference/android/net/ConnectivityManager#unregisterNetworkCallback(android.net.ConnectivityManager.NetworkCallback)):

> Notifications that would have triggered that NetworkCallback will immediately stop triggering it as soon as this call returns.

Reply:
> This PR achieves a simple Mutex I think by only using connlibSessionPtr if it's not null.

Not exactly, what can still happen is:

1. `disconnect` is called
2. While that's executing this function is called, we enter this callback, at that point `connlibSessionPtr` must not be null.
3. at some point of this function `disconnect` finish executing while some call into connlib is happening, such as `set_dns`
4. This is UB

Part of the problem could be handled with some kind of reference counting in rust-side. But after `disconnect` ends, `connlibSessionPtr` points to freed memory and the `setDns` would be called and `connlibSessionPtr` is not null.

Reply:
@conectado You're completely right -- I hadn't given that possibility much weight, but since we're in `unsafe` land it would be best to prevent that case from happening.

I'll wrap those calls in a mutex.

Reply:
Ok -- so I think this was being hit because we were calling `disconnect(it)` before `stopNetworkMonitor`.

I moved that around.

However, when connlib calls our `onDisconnect`, is the session pointer already freed at that point? If not, there's a race condition the client apps can't really do much about.